integrations/erc20/components/erc20-deploy.tsx

```diff
@@ -1,48 +1,52 @@
 
 import { WalletConnect } from '@/components/blockchain/wallet-connect'
 import { BranchIsWalletConnected } from '@/components/shared/branch-is-wallet-connected'
+import { Button } from '@/components/ui/button'
+import { Form, FormControl, FormDescription, FormField, FormItem, FormLabel, FormMessage } from '@/components/ui/form'
+import { Input } from '@/components/ui/input'
 
-import { erc20MintableABI } from '../abis/erc20-mintable-abi'
-import { erc20MintableByteCode } from '../abis/erc20-mintable-bytecode'
-import { useTokenStorage } from '../hooks/use-token-storage'
-
-export function DeployERC20Contract() {
-  const [token, setToken] = useTokenStorage()
-  const {
-    register,
-    handleSubmit,
-    formState: { errors },
-  } = useForm()
-  const { data: signer } = useSigner()
+import { useDeploy } from '../hooks/use-deploy'
+import { deployControls } from '../utils/controls'
 
-  const [, setContractAddress] = useState<string | undefined>()
-  const onSubmit = async (data: any) => {
-    // https://docs.ethers.org/v5/api/contract/example/#example-erc-20-contract--deploying-a-contract
-    const factory = new ethers.ContractFactory(erc20MintableABI, erc20MintableByteCode, signer as Signer)
-    const contract = await factory.deploy(data.name, data.symbol)
-    const deployed = await contract.deployTransaction.wait()
+const deployFormSchema = z.object({
+  name: z.string().min(2).max(50),
+  symbol: z.string().min(2).max(10),
+})
 
-    setToken(deployed.contractAddress)
-    setContractAddress(deployed.contractAddress)
-  }
+export function DeployERC20Contract() {
+  const { form, onSubmit, token } = useDeploy({ deployFormSchema })
 
   return (
     <>
-      <form className="flex flex-col gap-4" onSubmit={handleSubmit(onSubmit)}>
-        <label>Name</label>
-        <input {...register('name')} className="input" />
-        <label>Symbol</label>
-        <input {...register('symbol')} className="input" />
-        {errors.exampleRequired && <span>This field is required</span>}
-        <button type="submit" className="btn btn-emerald">
-          Deploy
-        </button>
-      </form>
+      <Form {...form}>
+        <form onSubmit={form.handleSubmit(onSubmit)} className="space-y-8">
+          {deployControls.map((item) => {
+            return (
+              <FormField
+                key={item?.label}
+                control={form.control}
+                name={item?.formfieldName}
+                render={({ field }) => (
+                  <>
+                    <FormItem>
+                      <FormLabel>{item?.label}</FormLabel>
+                      <FormControl>
+                        <Input placeholder={item?.placeholder} {...field} />
+                      </FormControl>
+                      <FormDescription>{item?.description}</FormDescription>
+                      <FormMessage />
+                    </FormItem>
+                  </>
+                )}
+              />
+            )
```

Comment:
you're return this exact same pattern on every form, so I suggest that we create a component for that, either for each form field or for multiple ones. Someting like that:
```
export function DefaultFormField(item, form) {
  return (
    <FormField
      key={item?.label}
      control={form.control}
      name={item?.formfieldName}
      render={({ field }) => (
        <>
          <FormItem>
            <FormLabel>{item?.label}</FormLabel>
            <FormControl>
              <Input placeholder={item?.placeholder} {...field} />
            </FormControl>
            <FormDescription>{item?.description}</FormDescription>
            <FormMessage />
          </FormItem>
        </>
      )}
    />
  )
}
```
or
```
export function DefaultFormFields(items, form) {
  {
    writeTransferControls.map((item) => {
      return (
        <FormField
          key={item?.label}
          control={form.control}
          name={item?.formfieldName}
          render={({ field }) => (
            <>
              <FormItem>
                <FormLabel>{item?.label}</FormLabel>
                <FormControl>
                  <Input placeholder={item?.placeholder} {...field} />
                </FormControl>
                <FormDescription>{item?.description}</FormDescription>
                <FormMessage />
              </FormItem>
            </>
          )}
        />
      )
    })
  }
}
```

Reply:
in the process of creating an <Item /> component which get its type from the controls and takes up its value from 
```
const getElement: { [key: string]: any } = {
  textArea: Textarea,
  input: Input,
  select: Select,
}

export function getComponent(key: string): any {
  return getElement[key]
}
```